Support multi-architecture Docker image

Juneezee opened the pull request:
As per #288, I have made this PR to make GitHub Actions supports building multi-architecture Docker image by using [docker/setup-qemu-action](https://github.com/docker/setup-qemu-action), [docker/setup-buildx-action](https://github.com/docker/setup-buildx-action), and [docker/build-push-action](https://github.com/docker/build-push-action).

I have tested the GitHub Actions workflow in my fork. The `docker` job depends on `build` job, and `docker` job will be run if and only if it is on `master` branch or a tag is pushed.

EDIT: This PR will make `.github/scripts/docker.bash` redundant, the maintainers can consider to remove it.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/bufbuild/buf?pullRequest=289) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/bufbuild/buf?pullRequest=289) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/bufbuild/buf?pullRequest=289) it.</sub>

ZymoticB commented on Dockerfile.buf:
```diff
@@ -7,14 +7,18 @@ RUN go mod download
+RUN CGO_ENABLED=0 GOOS=${TARGETOS} GOARCH=${TARGETARCH} \
+  go build -ldflags "-s -w" -trimpath -o /go/bin/buf ./cmd/buf
+
+FROM --platform=${BUILDPLATFORM} alpine:3.12.4
```
If I'm understanding how the buildx command works correctly, this final stage should be from the target platform

```suggestion
FROM --platform=${TARGETPLATFORM} alpine:3.12.4
```

Juneezee replied:
According to [Docker Buildx | Docker Documentation](https://docs.docker.com/buildx/working-with-buildx/#build-multi-platform-images)

> Finally, depending on your project, the language that you use may have good support for cross-compilation. In that case, multi-stage builds in Dockerfiles can be effectively used to build binaries for the platform specified with --platform using the native architecture of the build node. A list of build arguments like BUILDPLATFORM and TARGETPLATFORM is available automatically inside your Dockerfile and can be leveraged by the processes running as part of your build.

```
FROM --platform=$BUILDPLATFORM golang:alpine AS build
ARG TARGETPLATFORM
ARG BUILDPLATFORM
RUN echo "I am running on $BUILDPLATFORM, building for $TARGETPLATFORM" > /log
FROM alpine
COPY --from=build /log /log
```

Juneezee replied:
Yes, you are right.

Reference:
1. https://docs.docker.com/engine/reference/builder/#from
2. https://docs.docker.com/buildx/working-with-buildx/#build-multi-platform-images

bufdev commented on .github/workflows/ci.yaml:
```diff
@@ -8,11 +8,13 @@ jobs:
     steps:
       - name: checkout
         uses: actions/checkout@v2
+
```
no newline

bufdev commented on .github/workflows/ci.yaml:
```diff
@@ -8,11 +8,13 @@ jobs:
         uses: actions/setup-go@v2
         with:
           go-version: 1.16.2
+
```
no newline

bufdev commented on .github/workflows/ci.yaml:
```diff
@@ -25,20 +27,51 @@ jobs:
           key: ${{ runner.os }}-buf-${{ hashFiles('**/go.sum', '/make/**') }}
           restore-keys: |
             ${{ runner.os }}-buf-
+
```
no newline

bufdev commented on .github/workflows/ci.yaml:
```diff
@@ -25,20 +27,51 @@ jobs:
-          DOCKER_TOKEN: ${{ secrets.DOCKER_TOKEN }}
-          DOCKER_LATEST_BRANCH: master
-          DOCKER_VERSION_TAG_PREFIX: v
+
```
no newline

bufdev commented on Dockerfile.buf:
```diff
@@ -7,14 +7,18 @@ RUN go mod download
-    ca-certificates \
-    git \
-    openssh-client && \
+  ca-certificates \
```
revert indent

bufdev commented on .github/workflows/ci.yaml:
```diff
@@ -25,20 +27,51 @@ jobs:
+        with:
+          # alpine image doesn't support linux/riscv64
+          platforms: linux/386,linux/amd64,linux/arm64,linux/arm/v7,linux/arm/v6,linux/ppc64le,linux/s390x
+
```
no newline

bufdev commented on .github/workflows/ci.yaml:
```diff
@@ -25,20 +27,51 @@ jobs:
+
+      - name: setup-docker-buildx
+        uses: docker/setup-buildx-action@v1
+
```
no newline

bufdev commented on .github/workflows/ci.yaml:
```diff
@@ -25,20 +27,51 @@ jobs:
+        with:
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_TOKEN }}
+
```
no newline

bufdev commented on .github/workflows/ci.yaml:
```diff
@@ -25,20 +27,51 @@ jobs:
+          elif echo ${GITHUB_REF} | grep ^refs/tags/v >/dev/null; then
+            echo ::set-output name=tag::"$(echo "${GITHUB_REF}" | sed "s/refs\/tags\/v//")"
+          fi
+
```
no newline

bufdev reviewed (commented):
Thanks! We'll consider this change and do our own investigation.

Commenting in places that existing style isn't maintained - please match existing style.

This change isn't compatible with the Makefiles, and they will need to be updated as well.

bufdev commented on .github/workflows/ci.yaml:
```diff
@@ -25,20 +27,51 @@ jobs:
+        run: |
+          if echo "${GITHUB_REF}" | grep ^refs/heads/master$ >/dev/null; then
+            echo ::set-output name=tag::latest
+          elif echo ${GITHUB_REF} | grep ^refs/tags/v >/dev/null; then
```
quotes

ZymoticB commented on .github/workflows/ci.yaml:
```diff
@@ -32,13 +32,38 @@ jobs:
-          DOCKER_TOKEN: ${{ secrets.DOCKER_TOKEN }}
-          DOCKER_LATEST_BRANCH: master
-          DOCKER_VERSION_TAG_PREFIX: v
+  docker:
```
I think we've lost the check to ensure this isn't run on forks?

ZymoticB replied:
https://github.com/bufbuild/buf/blob/master/.github/scripts/docker.bash#L20

Juneezee replied:
I will update the `if` conditional to `github.head_ref == "" && (github.ref == 'refs/heads/master' || contains(github.ref, 'refs/tags/v'))`

ZymoticB replied:
I think the existing logic was actually incorrect. https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context

`github.head_ref` is only set during a PR, so a push to master / a version tag in a fork should pass this conditional (which we do not want). I believe the correct check (for skipping forks; the master / tags/v checks should remain) is 
```
github.repository == 'bufbuild/buf'
```

Additionally, can you add a comment with some of the details here so that a reader unexperienced with github actions understands the intent.

Juneezee replied:
I think you are right. I am not too sure what would be the output of `github.head_ref` actually. I was just using the logic from https://github.com/bufbuild/buf/blob/master/.github/scripts/docker.bash#L20 directly.

ZymoticB reviewed (commented):
Last round of comments I think, thanks for the contribution!

Juneezee commented on .github/workflows/ci.yaml:
```diff
@@ -32,13 +32,38 @@ jobs:
+      - name: set-tag
+        id: tag
+        run: |
+          if echo "${GITHUB_REF}" | grep ^refs/heads/master$ >/dev/null; then
```
Added the comment. Thanks for reviewing this PR :smiling_face_with_three_hearts:

ZymoticB commented:
@Juneezee it looks like https://github.com/bufbuild/buf/pull/289#discussion_r596237794 was not addressed but marked as resolved.

Juneezee commented:
> @Juneezee it looks like [#289 (comment)](https://github.com/bufbuild/buf/pull/289#discussion_r596237794) was not addressed but marked as resolved.

Sorry for missing that one. I have updated accordingly and added some comments.

ZymoticB reviewed (approved):
Awesome! This looks good to me. I'll coordinate with @bufdev to get it merged.
